`plugin-typescript` 12.1.3 cannot find modules with custom resolution logic

<!--
  ⚡️ katchow! We 💛 issues.

  🚨 Your issue will be CLOSED if:
     - This template is removed
     - Parts of this template are removed

  👉🏽 Need help or tech support? Please don't open an issue!
  Head to https://is.gd/rollup_chat or https://stackoverflow.com/questions/tagged/rollupjs

  ❤️ Rollup? Please consider supporting our collective:
  👉 https://opencollective.com/rollup/donate
-->

- Rollup Plugin Name: Typescript
- Rollup Plugin Version: 12.1.3
- Rollup Version: 4.43.0
- Operating System (or Browser): Linux 6.15.1-arch1-2, 
- Node Version: 24.2.0
- Link to reproduction _(⚠️ read below)_: https://stackblitz.com/edit/rollup-repro-sbbqvy8r?file=rollup.config.js

<!--
  🚨 Issues WITHOUT a valid reproduction WILL BE CLOSED!

  Please provide one by:
  1. Using the StackBlitz reproduction template at https://stackblitz.com/fork/rollup-repro
  2. Using the REPL.it plugin reproduction template at https://repl.it/@rollup/rollup-plugin-repro
  3. Provide a minimal repository link (Read https://gist.github.com/Rich-Harris/88c5fc2ac6dc941b22e7996af05d70ff for instructions).
     Please use NPM for installing dependencies!
     These may take more time to triage than the other options.

  ⚠️ ZIP Files are unsafe and maintainers will NOT download them.
-->


### Expected Behavior
On `plugin-typescript` 12.1.2 and below, it was possible to use a plugin with `resolveId` and `declare module "xyz" { ... }` to load modules outside the standard resolution path (for instance, loading generated sources from elsewhere).

For instance, with the following in your `rollup.config.js`.

```javascript
    plugins: [
        typescript(),
        {
            name: "custom",
            async resolveId(source) {
                if (source === "test-module") return path.resolve("generated/test-module.js")
            }
        }
    ]
```

and the following declaration file:

```typescript
declare module "test-module" {
    export const sayHello: () => void;
}
```

Then it was possible to import `"test-module"` from a Typescript file.

### Actual Behavior
This no longer works under the recent 12.1.3 release (I assume this is introduced by #1653), failing with the following:

```
(!) [plugin typescript] src/index.ts (1:26): @rollup/plugin-typescript TS2307: Cannot find module 'test-module' or its corresponding type declarations.
rollup-demo/src/index.ts:1:26

1 import { sayHello } from "test-module";
                           ~~~~~~~~~~~~~
```

### Additional Information

<!--
  Most issues can be expressed or demonstrated through the REPL or a repository.
  However, the situation may arise where some small code snippets also need to
  be provided. In that situation, please add your code below using
  Fenced Code Blocks (https://help.github.com/articles/creating-and-highlighting-code-blocks/)
-->

It's possible I'm doing something utterly unspeakable and unsupported! I wasn't sure if this was an accidental breaking change or just Hyrum's law in action, so felt worth reporting.

--- a/packages/typescript/src/index.ts
+++ b/packages/typescript/src/index.ts
@@ -2,7 +2,7 @@ import * as path from 'path';
 
 import { createFilter } from '@rollup/pluginutils';
 
-import type { Plugin, PluginContext, SourceDescription } from 'rollup';
+import type { Plugin, SourceDescription } from 'rollup';
 import type { Watch } from 'typescript';
 
 import type { RollupTypescriptOptions } from '../types';
@@ -37,31 +37,6 @@ export default function typescript(options: RollupTypescriptOptions = {}): Plugi
     tslib,
     typescript: ts
   } = getPluginOptions(options);
-  const createProgram = (context: PluginContext) =>
-    createWatchProgram(ts, context, {
-      formatHost,
-      resolveModule,
-      parsedOptions,
-      writeFile(fileName, data, _writeByteOrderMark, _onError, sourceFiles) {
-        if (sourceFiles) {
-          for (const sourceFile of sourceFiles) {
-            if (!parsedOptions.fileNames.includes(sourceFile.fileName)) {
-              parsedOptions.fileNames.push(sourceFile.fileName);
-            }
-          }
-        }
-
-        if (parsedOptions.options.composite || parsedOptions.options.incremental) {
-          tsCache.cacheCode(fileName, data);
-        }
-        emittedFiles.set(fileName, data);
-      },
-      status(diagnostic) {
-        watchProgramHelper.handleStatus(diagnostic);
-      },
-      transformers
-    });
-
   const tsCache = new TSCache(cacheDir);
   const emittedFiles = new Map<string, string>();
   const watchProgramHelper = new WatchProgramHelper();
@@ -81,14 +56,6 @@ export default function typescript(options: RollupTypescriptOptions = {}): Plugi
     name: 'typescript',
 
     buildStart(rollupOptions) {
-      if (typeof rollupOptions.input === 'string') {
-        rollupOptions.input = [rollupOptions.input];
-      }
-
-      if (Array.isArray(rollupOptions.input)) {
-        parsedOptions.fileNames = rollupOptions.input.map((fileName) => path.resolve(fileName));
-      }
-
       emitParsedOptionsErrors(ts, this, parsedOptions);
 
       preflight({
@@ -107,7 +74,21 @@ export default function typescript(options: RollupTypescriptOptions = {}): Plugi
         program = null;
       }
       if (!program) {
-        program = createProgram(this);
+        program = createWatchProgram(ts, this, {
+          formatHost,
+          resolveModule,
+          parsedOptions,
+          writeFile(fileName, data) {
+            if (parsedOptions.options.composite || parsedOptions.options.incremental) {
+              tsCache.cacheCode(fileName, data);
+            }
+            emittedFiles.set(fileName, data);
+          },
+          status(diagnostic) {
+            watchProgramHelper.handleStatus(diagnostic);
+          },
+          transformers
+        });
       }
     },
 
@@ -158,6 +139,7 @@ export default function typescript(options: RollupTypescriptOptions = {}): Plugi
 
       if (resolved) {
         if (/\.d\.[cm]?ts/.test(resolved.extension)) return null;
+        if (!filter(resolved.resolvedFileName)) return null;
         return path.normalize(resolved.resolvedFileName);
       }
 
@@ -167,20 +149,16 @@ export default function typescript(options: RollupTypescriptOptions = {}): Plugi
     async load(id) {
       if (!filter(id)) return null;
 
-      const resolvedId = path.resolve(id);
-
-      this.addWatchFile(resolvedId);
+      this.addWatchFile(id);
       await watchProgramHelper.wait();
 
-      const fileName = normalizePath(resolvedId);
+      const fileName = normalizePath(id);
       if (!parsedOptions.fileNames.includes(fileName)) {
         // Discovered new file that was not known when originally parsing the TypeScript config
-        parsedOptions.fileNames.push(path.resolve(fileName));
-
-        createProgram(this).close();
+        parsedOptions.fileNames.push(fileName);
       }
 
-      const output = findTypescriptOutput(ts, parsedOptions, resolvedId, emittedFiles, tsCache);
+      const output = findTypescriptOutput(ts, parsedOptions, id, emittedFiles, tsCache);
 
       return output.code != null ? (output as SourceDescription) : null;
     },

--- a/packages/typescript/test/fixtures/included-declarations/declarations.d.ts
+++ b/packages/typescript/test/fixtures/included-declarations/declarations.d.ts
@@ -1,0 +1,3 @@
+declare module 'declared-module' {
+  export function declaredFunction(): void;
+}

--- a/packages/typescript/test/fixtures/included-declarations/main.ts
+++ b/packages/typescript/test/fixtures/included-declarations/main.ts
@@ -1,0 +1,3 @@
+import { declaredFunction } from 'declared-module';
+
+declaredFunction();

--- a/packages/typescript/test/fixtures/included-declarations/tsconfig.json
+++ b/packages/typescript/test/fixtures/included-declarations/tsconfig.json
@@ -1,0 +1,3 @@
+{
+  "include": ["main.ts", "declarations.d.ts"],
+}

--- a/packages/typescript/test/fixtures/with-invalid-sources-inside-cwd/invalid.ts
+++ b/packages/typescript/test/fixtures/with-invalid-sources-inside-cwd/invalid.ts
@@ -1,2 +1,0 @@
-// eslint-disable-next-line
-foo

--- a/packages/typescript/test/fixtures/with-invalid-sources-inside-cwd/main.ts
+++ b/packages/typescript/test/fixtures/with-invalid-sources-inside-cwd/main.ts
@@ -1,3 +1,0 @@
-import {foo} from "./valid";
-
-console.log(foo);

--- a/packages/typescript/test/fixtures/with-invalid-sources-inside-cwd/valid.ts
+++ b/packages/typescript/test/fixtures/with-invalid-sources-inside-cwd/valid.ts
@@ -1,1 +1,0 @@
-export const foo = 5;

--- a/packages/typescript/test/test.js
+++ b/packages/typescript/test/test.js
@@ -1154,8 +1154,7 @@ test.serial('does it support tsconfig.rootDir for filtering', async (t) => {
   t.is(files.length, 1);
 });
 
-// todo: why would want to deliberately forbid resolution from outside of CWD? What problem does it solve to add such a constraint?
-test.skip('does it fail for filtering with incorrect rootDir in nested projects', async (t) => {
+test.serial('does it fail for filtering with incorrect rootDir in nested projects', async (t) => {
   process.chdir('fixtures/root-dir/packages/test-2');
   const error = await t.throwsAsync(
     rollup({
@@ -1615,17 +1614,18 @@ test.serial('compiled external library', async (t) => {
   t.pass();
 });
 
-test.serial(
-  'do not consider files that are not part of the entry point dependency graph',
-  async (t) => {
-    process.chdir('fixtures/with-invalid-sources-inside-cwd');
-    const input = 'main.ts';
+test.serial('observes included declarations', async (t) => {
+  const warnings = [];
+  const bundle = await rollup({
+    input: 'fixtures/included-declarations/main.ts',
+    external: ['declared-module'],
+    plugins: [typescript({ tsconfig: 'fixtures/included-declarations/tsconfig.json' })],
+    onwarn(warning) {
+      warnings.push(warning);
+    }
+  });
+  t.deepEqual(warnings, []);
 
-    const build = await rollup({
-      input,
-      plugins: [typescript()]
-    });
-
-    t.deepEqual(build.watchFiles, [path.resolve('main.ts'), path.resolve('valid.ts')]);
-  }
-);
+  const files = await getCode(bundle, { format: 'es' }, true);
+  t.is(files.length, 1);
+});

--- a/packages/typescript/test/tslib.ts
+++ b/packages/typescript/test/tslib.ts
@@ -1,5 +1,4 @@
 import { platform } from 'os';
-import { resolve } from 'path';
 
 import test from 'ava';
 import type { RollupError } from 'rollup';
@@ -63,15 +62,13 @@ test.serial('fails on bad tslib path', async (t) => {
     return;
   }
 
-  t.deepEqual(
-    error.message,
-    `Could not load fixtures/joker/tslib.js (imported by fixtures/overriding-tslib/main.ts): ENOENT: no such file or directory, open 'fixtures/joker/tslib.js'`
-  );
-  t.deepEqual(error.watchFiles, [
-    resolve('fixtures/overriding-tslib/main.ts'),
-    'fixtures/joker/tslib.js'
-  ]);
-  t.deepEqual(error.code, 'ENOENT');
+  if (error.watchFiles) {
+    let [filePath] = error.watchFiles;
+    filePath = filePath.substring(filePath.indexOf('packages'));
+    error.watchFiles[0] = filePath;
+  }
+
+  t.snapshot(error);
 });
 
 test.serial('fails without tslib installed', async (t) => {

